src/parser/parser.ee.ts

```diff
@@ -98,3 +98,225 @@ export async function parseStream({
+ * @param paragraphs The paragraph node(s) to process
+ * @returns Text extracted from all paragraphs
+ */
+function parseParagraphs(paragraphs: XMLNode[]): string {
+  let fullText = '';
+  for (const paragraph of paragraphs) {
+    const paragraphText = extractParagraphText(paragraph);
+    if (paragraphText.trim()) {
+      fullText += (fullText ? '\n' : '') + paragraphText;
+    }
+  }
+  return fullText.trim();
+}
+
+/**
+ * Extract text from a single paragraph node
+ * @param node The paragraph node
+ * @returns Text content of the paragraph
+ */
+function extractParagraphText(node: XMLNode): string {
+  let paragraphText = '';
+
+  /**
+   * Collect text recursively from a node
+   * @param node The current node
+   */
+  function collectText(node: XMLNode): void {
+    if (Array.isArray(node)) {
+      for (const child of node) {
+        collectText(child);
+      }
+    } else if (typeof node === 'object') {
+      if ('t' in node) {
+        const textContent = node.t;
+        if (typeof textContent === 'string') {
+          paragraphText = appendTextWithSpace(paragraphText, textContent);
+        } else if (typeof textContent === 'object' && '#text' in textContent) {
+          const text = textContent['#text'];
+          if (typeof text === 'string') {
+            paragraphText = appendTextWithSpace(paragraphText, text);
+          }
+        }
+      } else {
+        for (const child of Object.values(node)) {
+          collectText(child);
+        }
+      }
+    }
+  }
+
+  if (Array.isArray(node)) {
+    for (const child of node) {
+      collectText(child);
+    }
+  } else if (typeof node === 'object') {
+    if ('r' in node) {
+      collectText(node.r);
+    } else if (Array.isArray(node.r)) {
+      for (const run of node.r) {
+        collectText(run);
+      }
+    }
+  }
+
+  if (!paragraphText.endsWith('\n')) {
+    paragraphText += '\n';
+  }
+
+  return paragraphText.trim();
+}
+
+/**
+ * Appends text to the paragraph, ensuring a space is added if necessary.
+ * @param paragraphText The current paragraph text.
+ * @param textContent The text to append (will be trimmed).
+ * @returns The updated paragraph text.
+ */
+function appendTextWithSpace(
+  paragraphText: string,
+  textContent: string,
+): string {
+  textContent = textContent.trim();
+  if (textContent.length === 0) return paragraphText;
+
+  if (paragraphText.length > 0 && !paragraphText.endsWith(' ')) {
+    paragraphText += ' ';
+  }
+  return paragraphText + textContent;
+}
+
+/**
+ * Format text by fixing sentence terminations
+ * @param text Extracted text
+ * @returns Text with properly formatted sentences
+ */
+function formatText(text: string): string {
+  const sentences = text.split('\n').filter((sentence) => {
+    return sentence.trim() !== '';
+  });
+
+  const formattedSentences = sentences.map((sentence) => {
+    const trimmed = sentence.trim();
+
+    if (!/[!.?]$/.test(trimmed)) {
```

Comment:
please add a doc when you are using regex

+ refer to my comment about reusing the code from ansearch

Reply:
additionally, it seems to be one of my pieces of code from Ansearch, and I realised this was not always the best way to fix the end of paragraphs

leave it as is for now, don't fix paragraphs